File: java/Medium/_173_BinarySearchTreeIterator.java
package Medium;

import java.util.ArrayList;
import java.util.List;

import CS.TreeNode;

public class _173_BinarySearchTreeIterator {
    class BSTIterator {
        List<Integer> list = new ArrayList<>();
        int i = -1;
        public BSTIterator(TreeNode root) {
            preorder(root);
        }

        public void preorder(TreeNode node){
            if(node == null) return;
            preorder(node.left);
            list.add(node.val);
            preorder(node.right);
        }
        
        public int next() {
            return list.get(++i);
        }
        
        public boolean hasNext() {
            return i+1 < list.size();
        }
    }
    
}

File: Python3/_173_BinarySearchTreeIterator.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class BSTIterator:
    def __init__(self, root: Optional[TreeNode]):
        self.ar = []
        self.idx = 0
        self.preOrder(root)

    def preOrder(self,node:TreeNode):
        if node is None:
            return
        self.preOrder(node.left)
        self.ar.append(node.val)
        self.preOrder(node.right)

    def next(self) -> int:
        res = self.ar[self.idx]
        self.idx+=1
        return res

    def hasNext(self) -> bool:
        return self.idx < len(self.ar)
        


# Your BSTIterator object will be instantiated and called as such:
root = TreeNode(1)
obj = BSTIterator(root)
param_1 = obj.next()
print(param_1)
param_2 = obj.hasNext()
print(param_2)


root = TreeNode(5)
obj = BSTIterator(root)
param_1 = obj.next()
print(param_1)
param_2 = obj.hasNext()
print(param_2)